### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
purchase_data.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

In [3]:
#Player Count - Total Number of Players

total_players = purchase_data["SN"].nunique()
total_players

576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
unique_item = purchase_data["Item Name"].nunique()
unique_item

179

In [5]:
average_price = purchase_data["Price"].mean()
average_price

3.050987179487176

In [6]:
total_purchases = purchase_data["Item ID"].count()
total_purchases

780

In [7]:
total_revenue = purchase_data["Price"].sum()

total_revenue

2379.77

In [8]:
purchasing_df = pd.DataFrame({"Number of Unique Items": [unique_item], 
                              "Average Price": [average_price], 
                              "Number of Purchases": [total_purchases],
                              "Total Revenue": [total_revenue]})

purchasing_df["Average Price"] = purchasing_df["Average Price"].map('${:,.2f}'.format)
purchasing_df["Total Revenue"] = purchasing_df["Total Revenue"].map('${:,.2f}'.format)
purchasing_df.head()

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed





## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
gender_player_df = purchase_data.groupby(["Gender"])

gender_player_df.head()

Purchase ID             SN  Age                 Gender  Item ID  \
0              0        Lisim78   20                   Male      108   
1              1    Lisovynya38   40                   Male      143   
2              2     Ithergue48   24                   Male       92   
3              3  Chamassasya86   24                   Male      100   
4              4      Iskosia90   23                   Male      131   
9              9    Chanosian48   35  Other / Non-Disclosed      136   
15            15      Lisassa64   21                 Female       98   
18            18      Reunasu60   22                 Female       82   
22            22   Siarithria38   38  Other / Non-Disclosed       24   
38            38       Reulae52   10                 Female      116   
41            41      Assosia88   20                 Female        7   
55            55      Phaelap26   25                 Female       84   
82            82     Haerithp41   16  Other / Non-Disclosed      160   
111          111       Sundim98   21  Other / Non-Disclosed       41   
228          228    Jiskirran77   20  Other / Non-Disclosed       80   

                                     Item Name  Price  
0    Extraction, Quickblade Of Trembling Hands   3.53  
1                            Frenzied Scimitar   1.56  
2                                 Final Critic   4.88  
3                                  Blindscythe   3.27  
4                                         Fury   1.44  
9                 Ghastly Adamantite Protector   3.58  
15                 Deadline, Voice Of Subtlety   2.89  
18                                     Nirvana   4.90  
22                               Warped Fetish   3.81  
38                     Renewed Skeletal Katana   4.18  
41                Thorn, Satchel of Dark Souls   1.33  
55                                  Arcane Gem   3.79  
82                                  Azurewrath   4.40  
111                                      Orbit   4.75  
228                                  Dreamsong   3.39

In [10]:
gender_count = gender_player_df["SN"].nunique()

gender_count

Gender
Female                    81
Male                     484
Other / Non-Disclosed     11
Name: SN, dtype: int64

In [11]:
gender_percentage = gender_count/total_players *100

gender_percentage

Gender
Female                   14.062500
Male                     84.027778
Other / Non-Disclosed     1.909722
Name: SN, dtype: float64

In [12]:
gender_analysis = pd.DataFrame({"Total Count": gender_count, "Percentage of Players": gender_percentage})

gender_analysis["Percentage of Players"] = gender_analysis["Percentage of Players"].map("{:.2f}%".format)

gender_analysis

Total Count Percentage of Players
Gender                                                  
Female                          81                14.06%
Male                           484                84.03%
Other / Non-Disclosed           11                 1.91%

In [13]:
# Purchasing Analysis (Gender)

 
#Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender

#Create a summary data frame to hold the results

#Optional: give the displayed data cleaner formatting

#Display the summary data frame

In [14]:
gender_purchase_count = gender_player_df["Item ID"].count()
gender_purchase_count

Gender
Female                   113
Male                     652
Other / Non-Disclosed     15
Name: Item ID, dtype: int64

In [15]:
gender_avg_purchase_price = round(gender_player_df["Price"].mean(),2)
gender_avg_purchase_price

Gender
Female                   3.20
Male                     3.02
Other / Non-Disclosed    3.35
Name: Price, dtype: float64

In [16]:
gender_total_purchase = round(gender_player_df["Price"].sum(),2)
gender_total_purchase

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [17]:
gender_purchase_person = round(gender_total_purchase/gender_count, 2)

gender_purchase_person

Gender
Female                   4.47
Male                     4.07
Other / Non-Disclosed    4.56
dtype: float64

In [18]:
gender_purchase_analysis_chart = pd.DataFrame({"Purchase Count": gender_purchase_count, 
                                        "Average Purchase Price": gender_avg_purchase_price,
                                        "Total Purchase Value": gender_total_purchase,
                                        "Avg Total Purchase per Person": gender_purchase_person})

gender_purchase_analysis_chart["Average Purchase Price"] = gender_purchase_analysis_chart["Average Purchase Price"].map('${:,.2f}'.format)
gender_purchase_analysis_chart["Total Purchase Value"] = gender_purchase_analysis_chart["Total Purchase Value"].map('${:,.2f}'.format)
gender_purchase_analysis_chart["Avg Total Purchase per Person"] = gender_purchase_analysis_chart["Avg Total Purchase per Person"].map('${:,.2f}'.format)

gender_purchase_analysis_chart

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [19]:
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

groups_labels = ["<10", "10-14", "15-19","20-24","25-29","30-34", "35-39", "40+"]

In [20]:
age_player_df = purchase_data.loc[:,["Gender", "SN", "Age"]].drop_duplicates()

age_player_df["Age Range"] = pd.cut(age_player_df["Age"], bins, labels=groups_labels)

age_player_df.head()

Gender             SN  Age Age Range
0   Male        Lisim78   20     20-24
1   Male    Lisovynya38   40       40+
2   Male     Ithergue48   24     20-24
3   Male  Chamassasya86   24     20-24
4   Male      Iskosia90   23     20-24

In [21]:
age_count = age_player_df["Age Range"].value_counts()

age_count

20-24    258
15-19    107
25-29     77
30-34     52
35-39     31
10-14     22
<10       17
40+       12
Name: Age Range, dtype: int64

In [22]:
age_percent = round(age_count/total_players * 100, 2)

age_percent

20-24    44.79
15-19    18.58
25-29    13.37
30-34     9.03
35-39     5.38
10-14     3.82
<10       2.95
40+       2.08
Name: Age Range, dtype: float64

In [23]:
age_chart = pd.DataFrame({"Total Count": age_count, "Percentage of Players": age_percent})

age_chart["Percentage of Players"] = age_chart["Percentage of Players"].map("{:.2f}%".format)

age_chart = age_chart.sort_index()

age_chart

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [24]:
purchase_data["Age Range"] = pd.cut(purchase_data["Age"], bins, labels=groups_labels)

purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [25]:
purchase_analysis_data = purchase_data.groupby("Age Range")

purchase_analysis_data.head(10)

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
446          446   Chanossast57    7  Female      119   
477          477     Mindossa76   41    Male      173   
480          480      Salilis27   40    Male       59   
557          557     Frichaya88   42    Male        8   
667          667      Inasuir29   40  Female       66   

                                     Item Name  Price Age Range  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
446  Stormbringer, Dark Blade of Ending Misery   4.32       <10  
477                        Stormfury Longsword   4.93       40+  
480              Lightning, Etcher of the King   4.23       40+  
557                   Purgatory, Gem of Regret   3.93       40+  
667                         Victor Iron Spikes   4.40       40+  

[80 rows x 8 columns]

In [26]:
age_purchase = purchase_analysis_data["Item ID"].count()
age_purchase

Age Range
<10       23
10-14     28
15-19    136
20-24    365
25-29    101
30-34     73
35-39     41
40+       13
Name: Item ID, dtype: int64

In [27]:
avg_age_purchase = round(purchase_analysis_data["Price"].mean(), 2)

avg_age_purchase

Age Range
<10      3.35
10-14    2.96
15-19    3.04
20-24    3.05
25-29    2.90
30-34    2.93
35-39    3.60
40+      2.94
Name: Price, dtype: float64

In [28]:
total_age_purchase = round(purchase_analysis_data["Price"].sum(), 2)

total_age_purchase

Age Range
<10        77.13
10-14      82.78
15-19     412.89
20-24    1114.06
25-29     293.00
30-34     214.00
35-39     147.67
40+        38.24
Name: Price, dtype: float64

In [29]:
avg_total_age_person = round(total_age_purchase/age_count, 2)

avg_total_age_person

<10      4.54
10-14    3.76
15-19    3.86
20-24    4.32
25-29    3.81
30-34    4.12
35-39    4.76
40+      3.19
dtype: float64

In [30]:
purchase_age_chart = pd.DataFrame({"Purchase Count": age_purchase, 
                                   "Average Purchase Price": avg_age_purchase, 
                                   "Total Purchase Value": total_age_purchase, 
                                   "Avg Total Purchase per Person": avg_total_age_person})

purchase_age_chart["Average Purchase Price"] = purchase_age_chart["Average Purchase Price"].map("${:,.2f}".format)
purchase_age_chart["Total Purchase Value"] = purchase_age_chart["Total Purchase Value"].map("${:,.2f}".format)
purchase_age_chart["Avg Total Purchase per Person"] = purchase_age_chart["Avg Total Purchase per Person"].map("${:,.2f}".format)

purchase_age_chart.head(10)

Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [31]:
spender_player_df = purchase_data.groupby("SN")

spender_player_df.head()

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male       92   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price Age Range  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
775                                       Wolf   3.54     20-24  
776                           Exiled Doomblade   1.63     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     20-24  
778                               Final Critic   4.19       <10  
779                                       Dawn   4.60     20-24  

[780 rows x 8 columns]

In [32]:
spender_count = spender_player_df["Price"].count()

spender_count.head()

SN
Adairialis76    1
Adastirin33     1
Aeda94          1
Aela59          1
Aelaria33       1
Name: Price, dtype: int64

In [33]:
avg_spend_purchase = round(spender_player_df["Price"].mean(), 2)

avg_spend_purchase.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [34]:
total_spend_purchase = round(spender_player_df["Price"].sum(), 2)

total_spend_purchase.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [35]:
spender_chart = pd.DataFrame({"Purchase Count": spender_count, 
                              "Average Purchase Price": avg_spend_purchase, 
                              "Total Purchase Value": total_spend_purchase})

spender_chart = spender_chart.sort_values(["Total Purchase Value"], ascending = False)

spender_chart["Average Purchase Price"] = spender_chart["Average Purchase Price"].map("${:,.2f}".format)
spender_chart["Total Purchase Value"] = spender_chart["Total Purchase Value"].map("${:,.2f}".format)

spender_chart.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [36]:
popular_items_df = purchase_data.groupby(["Item ID", "Item Name"])

popular_items_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0              0        Lisim78   20   Male      108   
1              1    Lisovynya38   40   Male      143   
2              2     Ithergue48   24   Male       92   
3              3  Chamassasya86   24   Male      100   
4              4      Iskosia90   23   Male      131   
..           ...            ...  ...    ...      ...   
764          764   Saedaiphos46   18   Male      113   
765          765        Irith83   18   Male      130   
766          766   Aelastirin39   23   Male       58   
777          777     Yathecal72   20   Male       67   
779          779      Ennrian78   24   Male       50   

                                     Item Name  Price Age Range  
0    Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                            Frenzied Scimitar   1.56       40+  
2                                 Final Critic   4.88     20-24  
3                                  Blindscythe   3.27     20-24  
4                                         Fury   1.44     20-24  
..                                         ...    ...       ...  
764                          Solitude's Reaver   4.07     15-19  
765                                      Alpha   2.07     15-19  
766          Freak's Bite, Favor of Holy Might   4.14     20-24  
777      Celeste, Incarnation of the Corrupted   3.46     20-24  
779                                       Dawn   4.60     20-24  

[690 rows x 8 columns]

In [37]:
item_count = popular_items_df["Price"].count()

item_count.head()

Item ID  Item Name         
0        Splinter              4
1        Crucifer              4
2        Verdict               6
3        Phantomlight          6
4        Bloodlord's Fetish    5
Name: Price, dtype: int64

In [38]:
popular_item = round(popular_items_df["Price"].mean(), 2)

popular_item.head()

Item ID  Item Name         
0        Splinter              1.28
1        Crucifer              2.94
2        Verdict               2.48
3        Phantomlight          2.49
4        Bloodlord's Fetish    1.70
Name: Price, dtype: float64

In [39]:
item_price_purchase = round(popular_items_df["Price"].sum(),2)

item_price_purchase.head()

Item ID  Item Name         
0        Splinter               5.12
1        Crucifer              11.77
2        Verdict               14.88
3        Phantomlight          14.94
4        Bloodlord's Fetish     8.50
Name: Price, dtype: float64

In [40]:
pop_item_chart = pd.DataFrame({"Purchase Count": item_count, 
                               "Item Price": popular_item, 
                               "Total Purchase Value": item_price_purchase})

pop_item_chart_sort = pop_item_chart.sort_values(["Purchase Count"],ascending = False)

pop_item_chart_sort["Item Price"] = pop_item_chart_sort["Item Price"].map("${:,.2f}".format)
pop_item_chart_sort["Total Purchase Value"] = pop_item_chart_sort["Total Purchase Value"].map("${:,.2f}".format)

pop_item_chart_sort.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items 

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [41]:
pop_item_chart = pop_item_chart.sort_values(["Total Purchase Value"], ascending = False)

pop_item_chart["Item Price"] = pop_item_chart["Item Price"].map("${:,.2f}".format)
pop_item_chart["Total Purchase Value"] = pop_item_chart["Total Purchase Value"].map("${:,.2f}".format)

pop_item_chart.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## written description of three observable trends based on the data


* 84% of the players of this dataset are male (484/576). 
* 20 to 24 year olds are the largest age range for the game (248 players).
* Based on the purchasing anakysis, the top 3 spending age groups are 20-24 (365 purchases); 15-19(136 purchases); and 25-29(101 purchases). However, it is important to note that the age groups of 30-34 and 35-39; although the total purpase count is lower, the average perchase per person is higher: $4.12 and $4.76.
